Integrate IPOPT with Drake

sammy-tri opened the pull request:
Add IPOPT at an external, and an IpoptSolver class to wrap it.

IPOPT still doesn't pass all of the IK tests in MATLAB, so this PR still doesn't turn on the building of the mex files unless SNOPT is present.  I'll open an issue for tuning the solver and enabling that once this is merged.

Also, the top level CMakeLists.txt still refers to the new external in the sammy-tri github account, which is why this is marked do not merge.  I don't think this significantly impedes review, however.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2505)
<!-- Reviewable:end -->

sammy-tri commented:
+@lucy-tri for feature review

<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> #### Integrate IPOPT with Drake
> 
> Add IPOPT at an external, and an IpoptSolver class to wrap it.
> 
> 
> 
> IPOPT still doesn't pass all of the IK tests in MATLAB, so this PR still doesn't turn on the building of the mex files unless SNOPT is present.  I'll open an issue for tuning the solver and enabling that once this is merged.
> 
> 
> 
> Also, the top level CMakeLists.txt still refers to the new external in the sammy-tri github account, which is why this is marked do not merge.  I don't think this significantly impedes review, however.
> 
> 
> 
</details>


---

Review status: 0 of 11 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-:-KJI6E_gQFPVh14ZDqTb:1733109986)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
+@david-german-tri for platform review (but I'll wait for @lucy-tri's review)

<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> +@lucy-tri for feature review
</details>


---

Review status: 0 of 11 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-:-KJLc-EZku5kcfOu7HrN:-632154738)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
Publishing part 1 of my review...

<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> +@david-german-tri for platform review (but I'll wait for @lucy-tri's review)
</details>


---

Reviewed 10 of 11 files at r1.
Review status: 10 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[CMakeLists.txt, line 195 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJLecaRdSIXNQwfv2Ah:-KJLdyZCCZedvLEigzKX:-1292114513) ([raw file](https://github.com/robotlocomotion/drake/blob/cab99da2fd3efb41a2e54b2068067c75aa5417e5/CMakeLists.txt#L195)):*
> ```CMake
> set(gurobi_GIT_TAG fbbdbbc0152ddb37df209d221aff37448b8ff29f)
> set(gurobi_IS_PUBLIC FALSE)
> set(ipopt_GIT_REPOSITORY http://github.com/sammy-tri/ipopt-mirror.git)
> ```

As @sammy-tri  noted, this reference to his github is temp. I'm marking this for future reviewers.

---

*[drake/solvers/MathematicalProgram.cpp, line 99 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJLftnb6V4KQYd1EcAc:-KJLftncruh5ZECXjDSE:-649586026) ([raw file](https://github.com/robotlocomotion/drake/blob/cab99da2fd3efb41a2e54b2068067c75aa5417e5/drake/solvers/MathematicalProgram.cpp#L99)):*
> ```cpp
> 
>  private:
>   drake::solvers::IpoptSolver ipopt_solver;
> ```

Why do you need to say drake::solvers:: ?  You don't use that for the other solvers, below. If it's for clarity, go ahead and update the lines below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-:-KJLnDwmaCm_ycxWnRbH:-1990552777)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 10 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/MathematicalProgram.cpp, line 99 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJLftnb6V4KQYd1EcAc:-KJLo3omzmlpi1MTybVO:-426175683) ([raw file](https://github.com/robotlocomotion/drake/blob/cab99da2fd3efb41a2e54b2068067c75aa5417e5/drake/solvers/MathematicalProgram.cpp#L99)):*
<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary>
> Why do you need to say drake::solvers:: ?  You don't use that for the other solvers, below. If it's for clarity, go ahead and update the lines below.
</details>
Sigh...   Because the drake/solvers directory is only partially in compliance with the new namespace convention...  Much of the C++ code in here is just in the "Drake" namespace (including MathematialProgram), and I tried to follow the convention of "drake" as the top level and then the path to the file as the next namespace levels.

I can't actually find where this convention is written down, which is confusing me.

I should probably just do a cleanup PR for the entire drake/solvers directory which fixes this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 10 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/MathematicalProgram.cpp, line 99 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJLftnb6V4KQYd1EcAc:-KJLqVhFX36ahzjAd0LW:-63504363) ([raw file](https://github.com/robotlocomotion/drake/blob/cab99da2fd3efb41a2e54b2068067c75aa5417e5/drake/solvers/MathematicalProgram.cpp#L99)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Sigh...   Because the drake/solvers directory is only partially in compliance with the new namespace convention...  Much of the C++ code in here is just in the "Drake" namespace (including MathematialProgram), and I tried to follow the convention of "drake" as the top level and then the path to the file as the next namespace levels.
> 
> I can't actually find where this convention is written down, which is confusing me.
> 
> I should probably just do a cleanup PR for the entire drake/solvers directory which fixes this.
</details>
There actually isn't such a convention (matching the directories to the namespaces). That makes the namespace nesting unreasonably deep. For example, we decided a few PRs back that code in `drake/common` can just be namespace `drake::` -- adding `common::` to that is just clutter.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 10 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/MathematicalProgram.cpp, line 99 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJLftnb6V4KQYd1EcAc:-KJLqgk8GBc-n_ItcSgu:-843062294) ([raw file](https://github.com/robotlocomotion/drake/blob/cab99da2fd3efb41a2e54b2068067c75aa5417e5/drake/solvers/MathematicalProgram.cpp#L99)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> There actually isn't such a convention (matching the directories to the namespaces). That makes the namespace nesting unreasonably deep. For example, we decided a few PRs back that code in `drake/common` can just be namespace `drake::` -- adding `common::` to that is just clutter.
</details>
f2f discussion with @jwnimmer-tri indicates that there is no such rule applied to Drake, and we're only following the less restrictive rules in the google C++ guide.  I guess I should change this PR to use the Drake namespace like the rest of the solvers/ directory.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
Amazing job creating this code. 

<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary>
> Publishing part 1 of my review...
</details>


---

Review status: 10 of 11 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[drake/solvers/IpoptSolver.cpp, line 94 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJMa2ck4qRkA02I0C4C:-KJMa2ck4qRkA02I0C4D:1645470882) ([raw file](https://github.com/robotlocomotion/drake/blob/4ffcb67b1321a10c438c06aeda1085b572ad1b20/drake/solvers/IpoptSolver.cpp#L94)):*
> ```cpp
> 
> /// @return number of gradient entries populated
> size_t EvaluateConstraint(
> ```

Need to have comments describing what this is doing. Maybe a header comment, or maybe above each "for loop" section. 
I'm also puzzled why you don't apply constraint bounds like in nloptSolver. I'd guess you are doing it right, I just don't understand enough to understand this function.

---

*[drake/solvers/IpoptSolver.cpp, line 140 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJMf0e6a35qYIVOWY1P:-KJMf0e7evBG6eYzHaz8:-1765806449) ([raw file](https://github.com/robotlocomotion/drake/blob/4ffcb67b1321a10c438c06aeda1085b572ad1b20/drake/solvers/IpoptSolver.cpp#L140)):*
> ```cpp
>   }
> 
>   bool is_x_equal(Index n, const Number* x_in) {
> ```

Having n be an Index type makes it sound like n is an index into x, but it seems to be something else... it seems to be the size of x, or it should be, but I don't know why. Can you rename it or provide a comment?  
Ahhh...  n is the number of elements in what x_in is pointing to.  I see that we often have "Index n" in Ipopt.  And n is often used in math/programming to mean number of elements.  I guess it means here the number of elements in x_in. 
I could really use comments that describe the params, for all functions. 

---

*[drake/solvers/IpoptSolver.cpp, line 225 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJMnBaSKk0Kloisw1du:-KJMnBaT3qAO8GXlYPAS:551390663) ([raw file](https://github.com/robotlocomotion/drake/blob/4ffcb67b1321a10c438c06aeda1085b572ad1b20/drake/solvers/IpoptSolver.cpp#L225)):*
> ```cpp
>     }
>     for (const auto& c : problem_->linear_equality_constraints()) {
>       constraint_idx += GetConstraintBounds(
> ```

I don't see that you do anything with constraint_idx, and I don't see any side effects - if I'm missing something, can you document it?  I don't see the effect of the 3 for loops. 

---

*[drake/solvers/IpoptSolver.cpp, line 244 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJMwvo8DvmXWcovfaeE:-KJMwvo8DvmXWcovfaeF:1047104211) ([raw file](https://github.com/robotlocomotion/drake/blob/4ffcb67b1321a10c438c06aeda1085b572ad1b20/drake/solvers/IpoptSolver.cpp#L244)):*
> ```cpp
> 
>     assert(!init_z);
>     assert(!init_lambda);
> ```

Can you provide a comment on why you expect these to be false?

---

*[drake/solvers/IpoptSolver.cpp, line 306 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJMorFle3oCYXTWEaIN:-KJMorFmUdCHQfC8Onxh:1467101577) ([raw file](https://github.com/robotlocomotion/drake/blob/4ffcb67b1321a10c438c06aeda1085b572ad1b20/drake/solvers/IpoptSolver.cpp#L306)):*
> ```cpp
>         constraint_idx += c.constraint()->num_constraints();
>       }
>       assert(grad_idx == nele_jac);
> ```

So the only point of those 3 for loops is to do this assert?  
Also, I don't see where you generate any output: shouldn't you be writing to values?  I see the api says that if iRow and jCol are non-null you should write out to values.

---

*[drake/solvers/IpoptSolver.cpp, line 365 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJN0bXJwm9CTsil1MTJ:-KJN0bXJwm9CTsil1MTK:-823533505) ([raw file](https://github.com/robotlocomotion/drake/blob/4ffcb67b1321a10c438c06aeda1085b572ad1b20/drake/solvers/IpoptSolver.cpp#L365)):*
> ```cpp
>     TaylorVecXd this_x;
> 
>     memcpy(cost_cache_->x.data(), x, n * sizeof(Number));
> ```

I like the idea of a cache, but these memcpy's seem dangerous, if you get the size wrong. 

---

*[drake/solvers/IpoptSolver.cpp, line 396 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJN2V16rkXZlfhhwVhI:-KJN2V16rkXZlfhhwVhJ:1116229810) ([raw file](https://github.com/robotlocomotion/drake/blob/4ffcb67b1321a10c438c06aeda1085b572ad1b20/drake/solvers/IpoptSolver.cpp#L396)):*
> ```cpp
> 
>     for (const auto& c : problem_->generic_constraints()) {
>       grad += EvaluateConstraint(xvec, *c.constraint(), c.variable_list(),
> ```

elsewhere in this file you write this as *(c.constraint()). It's inconsistent. 

---

*[drake/solvers/MathematicalProgram.cpp, line 99 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJLftnb6V4KQYd1EcAc:-KJMQ4zjNgvJQmg5GpZo:-1380463375) ([raw file](https://github.com/robotlocomotion/drake/blob/cab99da2fd3efb41a2e54b2068067c75aa5417e5/drake/solvers/MathematicalProgram.cpp#L99)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> f2f discussion with @jwnimmer-tri indicates that there is no such rule applied to Drake, and we're only following the less restrictive rules in the google C++ guide.  I guess I should change this PR to use the Drake namespace like the rest of the solvers/ directory.
</details>
Ya, sounds good to eliminate the solvers namespace here. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-:-KJN3PMpPjXqWmhbIOQi:-1536887483)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/solvers/IpoptSolver.cpp, line 94 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJMa2ck4qRkA02I0C4C:-KJaDwQ8NJUVcGqK7yni:478563528) ([raw file](https://github.com/robotlocomotion/drake/blob/d48914fd5d9fe27255bce51e75c3a062be01a46b/drake/solvers/IpoptSolver.cpp#L94)):*
<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary>
> Need to have comments describing what this is doing. Maybe a header comment, or maybe above each "for loop" section. 
> I'm also puzzled why you don't apply constraint bounds like in nloptSolver. I'd guess you are doing it right, I just don't understand enough to understand this function.
</details>
Added some additional comments.  

---

*[drake/solvers/IpoptSolver.cpp, line 140 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJMf0e6a35qYIVOWY1P:-KJaEiqPLs8Vz1XnkdEF:259764873) ([raw file](https://github.com/robotlocomotion/drake/blob/d48914fd5d9fe27255bce51e75c3a062be01a46b/drake/solvers/IpoptSolver.cpp#L140)):*
<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary>
> Having n be an Index type makes it sound like n is an index into x, but it seems to be something else... it seems to be the size of x, or it should be, but I don't know why. Can you rename it or provide a comment?  
> Ahhh...  n is the number of elements in what x_in is pointing to.  I see that we often have "Index n" in Ipopt.  And n is often used in math/programming to mean number of elements.  I guess it means here the number of elements in x_in. 
> I could really use comments that describe the params, for all functions. 
</details>
Added additional comment here.

---

*[drake/solvers/IpoptSolver.cpp, line 225 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJMnBaSKk0Kloisw1du:-KJaFZD4mPMQufIe2Gt6:715257116) ([raw file](https://github.com/robotlocomotion/drake/blob/d48914fd5d9fe27255bce51e75c3a062be01a46b/drake/solvers/IpoptSolver.cpp#L225)):*
<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary>
> I don't see that you do anything with constraint_idx, and I don't see any side effects - if I'm missing something, can you document it?  I don't see the effect of the 3 for loops. 
</details>
Added some additional commentary clarifying that g_l and g_u are output variables and that constraint_idx is the offset into those arrays. 

---

*[drake/solvers/IpoptSolver.cpp, line 244 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJMwvo8DvmXWcovfaeE:-KJaGIdCjY6vx0XWAMBM:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d48914fd5d9fe27255bce51e75c3a062be01a46b/drake/solvers/IpoptSolver.cpp#L244)):*
<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary>
> Can you provide a comment on why you expect these to be false?
</details>
Done.

---

*[drake/solvers/IpoptSolver.cpp, line 306 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJMorFle3oCYXTWEaIN:-KJaGxFazPcntRkBiRIp:-2012836607) ([raw file](https://github.com/robotlocomotion/drake/blob/d48914fd5d9fe27255bce51e75c3a062be01a46b/drake/solvers/IpoptSolver.cpp#L306)):*
<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary>
> So the only point of those 3 for loops is to do this assert?  
> Also, I don't see where you generate any output: shouldn't you be writing to values?  I see the api says that if iRow and jCol are non-null you should write out to values.
</details>
Added some additional comments clarifying that iRow and jCol are output variables and the use of the index variables.  I read the API documentation as saying that either of values or (iRow, jCol) will be nullptr, and that all 3 will never be valid pointers.  The asserts in the function seems to back this up, so I don't think I'd ever want to do anything to values in the case that iRow and jCol are populated (not that I'm sure what this function would do with a null pointer anyway)

---

*[drake/solvers/IpoptSolver.cpp, line 365 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJN0bXJwm9CTsil1MTJ:-KJaId3ck10ZuhZFPS_h:-1524459767) ([raw file](https://github.com/robotlocomotion/drake/blob/d48914fd5d9fe27255bce51e75c3a062be01a46b/drake/solvers/IpoptSolver.cpp#L365)):*
<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary>
> I like the idea of a cache, but these memcpy's seem dangerous, if you get the size wrong. 
</details>
IPOPT's entire interface basically consists of scribbling at a memory location it hands you and not getting the size wrong...  I'm not sure this makes it much worse?

---

*[drake/solvers/IpoptSolver.cpp, line 396 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJN2V16rkXZlfhhwVhI:-KJaIyM0aS99gjk51slp:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d48914fd5d9fe27255bce51e75c3a062be01a46b/drake/solvers/IpoptSolver.cpp#L396)):*
<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary>
> elsewhere in this file you write this as *(c.constraint()). It's inconsistent. 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary>
> Amazing job creating this code. 
</details>


---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-:-KJaPabpDJHMVm4y45j5:1417088721)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/solvers/MathematicalProgram.cpp, line 99 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJLftnb6V4KQYd1EcAc:-KJaeiWUSRFCwW-Rhu_B:-754596165) ([raw file](https://github.com/robotlocomotion/drake/blob/cab99da2fd3efb41a2e54b2068067c75aa5417e5/drake/solvers/MathematicalProgram.cpp#L99)):*
<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary>
> Ya, sounds good to eliminate the solvers namespace here. 
</details>
From talking to @ggould-tri, I think we'd leaning toward going the other way, and (for the solvers/ directory, which is the extent of our control) going with drake::solvers for items linked into libdrakeOptimization.  (or, as @ggould-tri said "My intuition is that things that are getting built into different libraries by different build stanzas should be in distinct namespaces.").  system identification is the current precedent for this.  I'd say that the non-compliant items are a defect not to be fixed in this PR, but should be cleaned up later.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/solvers/MathematicalProgram.cpp, line 99 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJLftnb6V4KQYd1EcAc:-KJagu5PGNmzbKqj5jSh:2092286035) ([raw file](https://github.com/robotlocomotion/drake/blob/cab99da2fd3efb41a2e54b2068067c75aa5417e5/drake/solvers/MathematicalProgram.cpp#L99)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> From talking to @ggould-tri, I think we'd leaning toward going the other way, and (for the solvers/ directory, which is the extent of our control) going with drake::solvers for items linked into libdrakeOptimization.  (or, as @ggould-tri said "My intuition is that things that are getting built into different libraries by different build stanzas should be in distinct namespaces.").  system identification is the current precedent for this.  I'd say that the non-compliant items are a defect not to be fixed in this PR, but should be cleaned up later.
</details>
OK, yes, I read that discussion on Slack too. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 10 of 11 files reviewed at latest revision, 22 unresolved discussions.

---

*[drake/solvers/CMakeLists.txt, line 66 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfGyDlH-9eOj-6Exvt:-KJfGyDlH-9eOj-6Exvu:-433289887) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/CMakeLists.txt#L66)):*
> ```CMake
>   if (NOT (APPLE OR WIN32))
>     set(CMAKE_SHARED_LINKER_FLAGS
>       "${CMAKE_SHARED_LINKER_FLAGS} -Wl,-Bsymbolic -Wl,-Bsymbolic-functions")
> ```

why is this needed?

---

*[drake/solvers/CMakeLists.txt, line 67 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfGnn1l9bA_8aapzgo:-KJfGnn2duMz58UojnXi:-1185040332) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/CMakeLists.txt#L67)):*
> ```CMake
>     set(CMAKE_SHARED_LINKER_FLAGS
>       "${CMAKE_SHARED_LINKER_FLAGS} -Wl,-Bsymbolic -Wl,-Bsymbolic-functions")
>   endif(NOT WIN32)
> ```

unbalanced `endif` (just `endif()` would be fine)

---

*[drake/solvers/IpoptSolver.cpp, line 2 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfHL3ml1C_klMS8Wej:-KJfHL3ml1C_klMS8Wek:358257291) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L2)):*
> ```cpp
> 
> #include "drake/solvers/IpoptSolver.h"
> ```

Is the plan still to do a later cleanup pass for file name style?  As I've said before, I think writing new code in style from the beginning would be preferable.  Your call, but I will continue mentioning it once per PR.

---

*[drake/solvers/IpoptSolver.cpp, line 38 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfIg1uLXchnpCNgyCG:-KJfIg1uLXchnpCNgyCH:-1356436703) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L38)):*
> ```cpp
>   const Eigen::VectorXd& upper_bound = c.upper_bound();
>   for (size_t i = 0; i < c.num_constraints(); i++) {
>     lb[i] = lower_bound(i);
> ```

BTW, this is terrifying but I guess that's IPOPT's fault.

---

*[drake/solvers/IpoptSolver.cpp, line 69 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfKBG5bD7e4egAjJXF:-KJfKBG5bD7e4egAjJXG:763814699) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L69)):*
> ```cpp
> ///
> /// @return the number of row/column pairs filled in.
> size_t GetGradientMatrix(
> ```

FYI, I let @jwnimmer-tri talk me out of using `size_t` because it's unsigned, but I was never happy with that outcome, and on closer examination of the style guide I noticed it says "you are welcome to use standard types like `size_t`".  So, I'm going to put it back in my code, and I approve of your usage here.

https://google.github.io/styleguide/cppguide.html#Integer_Types

---

*[drake/solvers/IpoptSolver.cpp, line 71 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfLr781G36-_y6Lh5s:-KJfLr781G36-_y6Lh5t:1824189469) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L71)):*
> ```cpp
> size_t GetGradientMatrix(
>     const Drake::Constraint& c, const Drake::VariableList& variable_list,
>     Index constraint_idx, Index* iRow, Index* jCol) {
> ```

Please document the `constraint_idx` parameter.

---

*[drake/solvers/IpoptSolver.cpp, line 76 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfLXcP1hKqEN99prIf:-KJfLPJMEaHbhinMqmWl:-2142046556) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L76)):*
> ```cpp
> 
>   for (const DecisionVariableView& v : variable_list) {
>     for (size_t i = 0; i < m; i++) {
> ```

FYI, this would be easier to read with fewer single letter variables.  Consider `num_constraints` and `decision_variable` - wouldn't even cause any line wrap!

---

*[drake/solvers/IpoptSolver.cpp, line 95 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfNI5BEXC3ycNgiyrF:-KJfNI5BEXC3ycNgiyrG:943374496) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L95)):*
> ```cpp
> }
> 
> /// @return number of gradient entries populated
> ```

Please explain what this function does!

---

*[drake/solvers/IpoptSolver.cpp, line 99 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfUi4Xeo1WOVn4sWOK:-KJfP_uIg97Ztlc1Yza2:-1250629103) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L99)):*
> ```cpp
>     const Eigen::VectorXd& xvec,
>     const Drake::Constraint& c, const Drake::VariableList& variable_list,
>     Number* result, Number* grad) {
> ```

Am I understanding correctly - grad is sparse, and its structure is determined by GetGradientMatrix?  If so, that would be good to capture in the function comment.

---

*[drake/solvers/IpoptSolver.cpp, line 132 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfNS5_Rg9nDm26DM0t:-KJfNS5_Rg9nDm26DM0u:1924878781) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L132)):*
> ```cpp
>   // Extract the appropriate derivatives from our result into the
>   // gradient array.  Like above, we need to use variable_list to
>   // figure out where the drivatives we actually care about are
> ```

typo: derivatives

---

*[drake/solvers/IpoptSolver.cpp, line 148 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfP7I80rQVCnPexWd9:-KJfP7I9cNQtarqgudTA:1193677426) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L148)):*
> ```cpp
> // Since Drake's eval() functions emit both of these at once, cache
> // the result for IPOPT.
> struct ResultCache {
> ```

Consider making this a class instead.

https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes

---

*[drake/solvers/IpoptSolver.cpp, line 150 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfOZbr1_NQokEZcliZ:-KJfOZbr1_NQokEZcli_:1666792674) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L150)):*
> ```cpp
> struct ResultCache {
>   ResultCache(size_t x_size, size_t result_size, size_t grad_size) {
>     // The choice of infinity as the default value below is arbitrary.
> ```

FYI, some people have argued for NaN initialization in other parts of the code base.  I don't particularly care, but it may have some advantages over infinity.

---

*[drake/solvers/IpoptSolver.cpp, line 182 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfPJCpvyj-zGCO7lhc:-KJfPJCpvyj-zGCO7lhd:-2125738146) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L182)):*
> ```cpp
>   virtual ~IpoptSolver_NLP() {}
> 
>   virtual bool get_nlp_info(Index& n, Index& m, Index& nnz_jac_g,
> ```

`override`?  here and below

---

*[drake/solvers/IpoptSolver.cpp, line 186 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfTSNLdU3412z_ZcXN:-KJfTSNLdU3412z_ZcXO:1568422888) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L186)):*
> ```cpp
>     n = problem_->num_vars();
> 
>     cost_cache_.reset(new ResultCache(n, 1, n));
> ```

please add a comment explaining why these are the right dimensions.

---

*[drake/solvers/IpoptSolver.cpp, line 273 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfTL0zzDJSscvY6B3x:-KJfTL0zzDJSscvY6B3y:1581370527) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L273)):*
> ```cpp
>     }
> 
>     obj_value = cost_cache_->result[0];
> ```

// cost_cache_->result is length 1.

---

*[drake/solvers/IpoptSolver.cpp, line 373 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfUwk3j9zmpvb2uA_w:-KJfUwk3j9zmpvb2uA_x:1637608787) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L373)):*
> ```cpp
>     }
> 
>     Eigen::VectorXd sol(n);
> ```

`solution`?

---

*[drake/solvers/IpoptSolver.cpp, line 443 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfQbgEG_73Pf4BS9hB:-KJfQbgFORnD9RqZ-AfU:148778315) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L443)):*
> ```cpp
> };
> 
> }  // end anonymous namespace
> ```

nit: just `// namespace`

https://google.github.io/styleguide/cppguide.html#Namespace_Formatting

---

*[drake/solvers/IpoptSolver.cpp, line 461 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfVepIEI0WsdyMFOkY:-KJfVepIEI0WsdyMFOkZ:-601688403) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L461)):*
> ```cpp
>   app->Options()->SetIntegerValue("print_level", 2);
> 
>   for (const auto it : prog.GetSolverOptionsDouble("IPOPT")) {
> ```

`const auto&`, here and below

---

*[drake/solvers/Optimization.h, line 674 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfVxwxdbZ6KamrIH1A:-KJfVxwxdbZ6KamrIH1B:-481815635) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/Optimization.h#L674)):*
> ```objc
>   void SetSolverOption(const std::string& solver_name,
>                        const std::string& solver_option,
>                        const std::string option_value) {
> ```

`&` also?

---

*[drake/solvers/test/testOptimizationProblem.cpp, line 208 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfWCxdIlET48BN51fV:-KJfWCxdIlET48BN51fW:917665670) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/test/testOptimizationProblem.cpp#L208)):*
> ```cpp
>   VectorXd expected(5);
>   expected << 1, 1, 0, 1, 0;
>   prog.SetInitialGuess({x}, expected + .01 * VectorXd::Random(5));
> ```

Same comment as below?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 7 of 11 files reviewed at latest revision, 18 unresolved discussions.

---

*[CMakeLists.txt, line 195 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJLecaRdSIXNQwfv2Ah:-KJfkbFtvWi-IBKqajvy:619094326) ([raw file](https://github.com/robotlocomotion/drake/blob/cab99da2fd3efb41a2e54b2068067c75aa5417e5/CMakeLists.txt#L195)):*
<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary>
> As @sammy-tri  noted, this reference to his github is temp. I'm marking this for future reviewers.
</details>
Leaving this marked as blocking until I get someone to move the code into RobotLocomotion.

---

*[drake/solvers/CMakeLists.txt, line 66 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfGyDlH-9eOj-6Exvt:-KJfg-m77veZzD3D2D3V:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/CMakeLists.txt#L66)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> why is this needed?
</details>
Done.

---

*[drake/solvers/CMakeLists.txt, line 67 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfGnn1l9bA_8aapzgo:-KJfg09D-n2ByqNBcsYN:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/CMakeLists.txt#L67)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> unbalanced `endif` (just `endif()` would be fine)
</details>
Done.

---

*[drake/solvers/IpoptSolver.cpp, line 2 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfHL3ml1C_klMS8Wej:-KJfg1dKJoXlH4HrRcni:461894799) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L2)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Is the plan still to do a later cleanup pass for file name style?  As I've said before, I think writing new code in style from the beginning would be preferable.  Your call, but I will continue mentioning it once per PR.
</details>
I'll be filing an issue to fix namespace and file names throughout drake/solvers/ after this PR lands.

---

*[drake/solvers/IpoptSolver.cpp, line 71 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfLr781G36-_y6Lh5s:-KJfgW0QtZXLEp1bSzNZ:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L71)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Please document the `constraint_idx` parameter.
</details>
Done.

---

*[drake/solvers/IpoptSolver.cpp, line 95 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfNI5BEXC3ycNgiyrF:-KJfh4AGf38C9RMt8tDM:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L95)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Please explain what this function does!
</details>
Done.

---

*[drake/solvers/IpoptSolver.cpp, line 99 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfUi4Xeo1WOVn4sWOK:-KJfhO4MU3Tyn8gJYrC_:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L99)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Am I understanding correctly - grad is sparse, and its structure is determined by GetGradientMatrix?  If so, that would be good to capture in the function comment.
</details>
Done.

---

*[drake/solvers/IpoptSolver.cpp, line 132 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfNS5_Rg9nDm26DM0t:-KJfhRrUAFAPVv71zRGD:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L132)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> typo: derivatives
</details>
Done.

---

*[drake/solvers/IpoptSolver.cpp, line 148 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfP7I80rQVCnPexWd9:-KJfhfuyAOk-W9SI4BPK:-794914953) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L148)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Consider making this a class instead.
> 
> https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes
</details>
I consulted that section when originally deciding struct vs. class for ResultCache.  IMO, while the equality test does exceed "functionality other than access/setting the data members," it wasn't by enough to matter.  If I were to rectify this, I'd be nearly as likely tp move the equality test to a free function.  At any rate, it's currently only a "Consider".

---

*[drake/solvers/IpoptSolver.cpp, line 182 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfPJCpvyj-zGCO7lhc:-KJfiNo5c-OpSb0HZMaA:1202263213) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L182)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> `override`?  here and below
</details>
" Older (pre-C++11) code will use the virtual keyword as an inferior alternative annotation. For clarity, use exactly one of override, final, or virtual when declaring an override"

My use of "virtual" here was to be consistent with the (pre-C++11) style in the IPOPT documentation: http://www.coin-or.org/Ipopt/documentation/node23.html

---

*[drake/solvers/IpoptSolver.cpp, line 186 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfTSNLdU3412z_ZcXN:-KJfjJAXhzbzef_RXEhs:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L186)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> please add a comment explaining why these are the right dimensions.
</details>
Done.

---

*[drake/solvers/IpoptSolver.cpp, line 273 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfTL0zzDJSscvY6B3x:-KJfjVxLQmCfn7I-Tq-O:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L273)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> // cost_cache_->result is length 1.
</details>
Done.

---

*[drake/solvers/IpoptSolver.cpp, line 373 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfUwk3j9zmpvb2uA_w:-KJfjak_I4oBm2ONzuYI:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L373)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> `solution`?
</details>
Done.

---

*[drake/solvers/IpoptSolver.cpp, line 443 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfQbgEG_73Pf4BS9hB:-KJfjdp4_uoIxqk5QFYo:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L443)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> nit: just `// namespace`
> 
> https://google.github.io/styleguide/cppguide.html#Namespace_Formatting
</details>
Done.

---

*[drake/solvers/IpoptSolver.cpp, line 461 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfVepIEI0WsdyMFOkY:-KJfjklsUoRb1cQrdS5l:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L461)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> `const auto&`, here and below
</details>
Done.

---

*[drake/solvers/Optimization.h, line 674 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfVxwxdbZ6KamrIH1A:-KJfjyjoFJE9U8eDtdlX:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/Optimization.h#L674)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> `&` also?
</details>
Done.

---

*[drake/solvers/test/testOptimizationProblem.cpp, line 208 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfWCxdIlET48BN51fV:-KJfkM_OC4MAdTYV9YXA:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/test/testOptimizationProblem.cpp#L208)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Same comment as below?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary>
> :lgtm:
</details>


---

Review status: 7 of 11 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/solvers/IpoptSolver.cpp, line 182 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-KJfPJCpvyj-zGCO7lhc:-KJfsNZS1_yi1pamdE7p:-1007049327) ([raw file](https://github.com/robotlocomotion/drake/blob/d5b2bce754ee2375d75a72bc4ce17bb07691b6b9/drake/solvers/IpoptSolver.cpp#L182)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> " Older (pre-C++11) code will use the virtual keyword as an inferior alternative annotation. For clarity, use exactly one of override, final, or virtual when declaring an override"
> 
> My use of "virtual" here was to be consistent with the (pre-C++11) style in the IPOPT documentation: http://www.coin-or.org/Ipopt/documentation/node23.html
</details>
OK - not the choice I would have made, but a good enough reason.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-:-KJfscBNhrJfR42nd-NE:1417088721)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
Reviewed 1 of 1 files at r4, 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Ok then, I think it's up to Russ (or anyone else with write privs to the RobotLocomotion github...  david, does this include you?) to move http://github.com/sammy-tri/ipopt-mirror.git out my sandbox and for me to update the URL.  Once that's done, I'll clear the "do not merge" tag.

@sherm1, can you acknowledge your comment if you're OK with the result of the discussion?

<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> :lgtm:
</details>


---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-:-KJg-w2_jlsY3oGiASpA:639987156)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
I do not have those bits.

<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Ok then, I think it's up to Russ (or anyone else with write privs to the RobotLocomotion github...  david, does this include you?) to move http://github.com/sammy-tri/ipopt-mirror.git out my sandbox and for me to update the URL.  Once that's done, I'll clear the "do not merge" tag.
> 
> @sherm1, can you acknowledge your comment if you're OK with the result of the discussion?
</details>


---

Reviewed 7 of 11 files at r1, 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-:-KJg2dAs_dNNU55euuxj:1830480195)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Acknowledged; not something to resolve in this PR.

<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> I do not have those bits.
</details>


---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-:-KJg94whDy44U7Mhmgxc:-1353599695)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-:-KKEzndzHxpqDg1jQGs2:1417088721)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r8.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2505#-:-KKFR_oPUZfEJC0ts8mc:1417088721)*
<!-- Sent from Reviewable.io -->
